contracts/contracts/coordination/Subscription.sol

```diff
@@ -0,0 +1,271 @@
+     * @return The base expiration duration
+     */
+    function baseExpiration() public pure returns (uint256) {
+        return 52 weeks; // TODO
+    }
+
+    /**
+     * @notice Creates a new subscription
+     * @param ritualId The ID of the ritual
+     */
+    function newSubscription(uint32 ritualId) external {
+        uint32 subscriptionId = numberOfSubscriptions;
+        SubscriptionInfo storage sub = subscriptions[subscriptionId];
+        sub.subscriber = msg.sender;
+        paySubscriptionFor(subscriptionId);
+
+        subscribers[LookupKey.lookupKey(ritualId, msg.sender)] = subscriptionId;
+        numberOfSubscriptions += 1;
+
+        emit SubscriptionCreated(subscriptionId, msg.sender, ritualId);
+    }
+
+    /**
+     * @notice Pays for a subscription
+     * @param subscriptionId The ID of the subscription
+     */
+    function paySubscriptionFor(uint32 subscriptionId) public virtual {
+        uint256 amount = subscriptionFee();
+
+        SubscriptionInfo storage sub = subscriptions[subscriptionId];
+        sub.paidFor += amount;
+        sub.expiration += baseExpiration();
+
+        feeToken.safeTransferFrom(msg.sender, address(this), amount);
+
+        // TODO: We already emit SubscriptionCreated, do we need this?
+        emit SubscriptionPaid(subscriptionId, msg.sender, amount);
+    }
+
+    /**
+     * @notice Checks if a spender can spend from a subscription
+     * @param subscriptionId The ID of the subscription
+     * @param spender The address of the spender
+     * @return True if the spender can spend from the subscription, false otherwise
+     */
+    function canSpendFromSubscription(
+        // TODO: Currently unused, remove?
+        // solhint-disable-next-line no-unused-vars
+        uint32 subscriptionId,
+        address spender
+    ) public returns (bool) {
+        // By default, only coordinator can spend from subscription
+        return spender == address(coordinator);
+    }
+
+    /**
+     * @notice Spends from a subscription
+     * @param subscriptionId The ID of the subscription
+     * @param amount The amount to spend
+     */
+    function spendFromSubscription(uint32 subscriptionId, uint256 amount) external {
+        require(canSpendFromSubscription(subscriptionId, msg.sender), "Unauthorized spender");
+        feeToken.safeTransferFrom(address(this), msg.sender, amount);
+    }
+
+    /**
+     * @notice Withdraws the contract balance to the beneficiary
+     * @param amount The amount to withdraw
+     */
+    function withdrawToBeneficiary(uint256 amount) external {
+        require(msg.sender == beneficiary, "Only the beneficiary can withdraw");
+
+        uint256 availableAmount = 0;
+        for (uint32 i = 0; i < numberOfSubscriptions; i++) {
+            SubscriptionInfo storage sub = subscriptions[i];
+            if (block.timestamp >= sub.expiration) {
+                availableAmount += sub.paidFor - sub.spent;
+            }
+        }
```

Comment:
sorry to nitpick again here, how about to extract this calculation to public view function? it will help to caller to check in advance how much can be withdrawn